packages/lit-html/src/directives/ref.ts

```diff
@@ -0,0 +1,133 @@
+/**
+ * @license
+ * Copyright (c) 2020 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+import {directive, nothing, ElementPart} from '../lit-html.js';
+import {DisconnectableDirective} from '../disconnectable-directive.js';
+
+/**
+ * Creates a new Ref object, which is container for a reference to an element.
+ */
+export const createRef = <T = Element>() => new Ref<T>();
+
+class Ref<T = Element> {
+  /**
+   * The current Element value of the ref, or else `undefined` if the ref is no
+   * longer rendered.
+   */
+  readonly value?: T;
+}
+
+interface RefInternal {
+  value: Element | undefined;
+}
+
+// When callbacks are used for refs, this map tracks the last value the callback
+// was called with, for ensuring a directive doesn't clear the ref if the ref
+// has already been rendered to a new spot
+const lastElementForCallback: WeakMap<
+  Function,
+  Element | undefined
+> = new WeakMap();
+
+export type RefOrCallback = Ref | ((el: Element | undefined) => void);
+
+class RefDirective extends DisconnectableDirective {
+  private _element?: Element;
+  private _ref?: RefOrCallback;
+
+  render(_ref: RefOrCallback) {
+    return nothing;
+  }
+
+  update(part: ElementPart, [ref]: Parameters<this['render']>) {
+    const refChanged = ref !== this._ref;
+    if (refChanged && this._ref !== undefined) {
+      // The ref passed to the directive has changed;
+      // unset the previous ref's value
+      this._updateRefValue(undefined);
+    }
+    if (refChanged || this._lastElementForRef !== this._element) {
+      // We either got a new ref or this is the first render;
+      // store the ref/element & update the ref value
+      this._ref = ref;
+      this._updateRefValue((this._element = part.element));
+    }
+    return nothing;
+  }
+
+  private _updateRefValue(element: Element | undefined) {
+    if (typeof this._ref === 'function') {
+      // If the current ref was called with a previous value,
+      // call with `undefined`
```

Comment:
I'd leave a little comment why we do this. There are cases where we can't _not_ call with undefined when the ref changes and this is to normalize all the cases, yeah?

Reply:
Exactly. I'll clarify.